pkg/openfeature/client.go

```diff
@@ -531,8 +573,8 @@ func (c Client) evaluate(
 		},
 	}
 
-	apiClientInvocationProviderHooks := append(append(append(api.hooks, c.hooks...), options.hooks...), api.provider.Hooks()...) // API, Client, Invocation, Provider
-	providerInvocationClientApiHooks := append(append(append(api.provider.Hooks(), options.hooks...), c.hooks...), api.hooks...) // Provider, Invocation, Client, API
+	apiClientInvocationProviderHooks := append(append(append(api.hooks(), c.hooks...), options.hooks...), api.provider().Hooks()...) // API, Client, Invocation, Provider
+	providerInvocationClientApiHooks := append(append(append(api.provider().Hooks(), options.hooks...), c.hooks...), api.hooks()...) // Provider, Invocation, Client, API
```

Comment:
I think I see one nasty problem here. If between this line and the evaluation, I switch providers, the hooks from the first provider will run with the evaluation of the second.

I think the only way to resolve it would be to expose a lock method on the API which could be read-locked locking at the point the hooks are retrieved, and unlocking after evaluation.

I don't know if adding that sort of locking is worth it, when you consider that providers are switched infrequently.

Reply:
Perhaps we could defer getting provider hooks until the lock is active? I know that would make a mess.

Reply:
Great observation, I've opted for the approach of locking a provider and its hooks to an evaluation transaction. Another thread is free to set a new provider but any ongoing transactions will carry on using the existing provider and its hooks.
If there are any concerns we can revise this solution.